File: packages/lua/queue.lua
local queue = {}

-- --- Queue ---
local Queue = {}
Queue.__index = Queue

function queue.New()
    return setmetatable({elements = {}}, Queue)
end

function Queue:Enqueue(value)
    table.insert(self.elements, value)
end

function Queue:Dequeue()
    if self:IsEmpty() then return nil, "queue is empty" end
    return table.remove(self.elements, 1)
end

function Queue:Peek()
    if self:IsEmpty() then return nil, "queue is empty" end
    return self.elements[1]
end

function Queue:IsEmpty()
    return #self.elements == 0
end

function Queue:Size()
    return #self.elements
end

-- --- Deque ---
local Deque = {}
Deque.__index = Deque

function queue.NewDeque()
    return setmetatable({elements = {}}, Deque)
end

function Deque:PushBack(value)
    table.insert(self.elements, value)
end

function Deque:PushFront(value)
    table.insert(self.elements, 1, value)
end

function Deque:PopBack()
    if #self.elements == 0 then return nil, "deque is empty" end
    return table.remove(self.elements)
end

function Deque:PopFront()
    if #self.elements == 0 then return nil, "deque is empty" end
    return table.remove(self.elements, 1)
end

-- --- Circular Queue ---
local CircularQueue = {}
CircularQueue.__index = CircularQueue

function queue.NewCircularQueue(capacity)
    return setmetatable({elements = {}, capacity = capacity, head = 1, tail = 1, size = 0}, CircularQueue)
end

function CircularQueue:Enqueue(value)
    if self:IsFull() then return false, "queue is full" end
    self.elements[self.tail] = value
    self.tail = (self.tail % self.capacity) + 1
    self.size = self.size + 1
    return true
end

function CircularQueue:Dequeue()
    if self:IsEmpty() then return nil, "queue is empty" end
    local val = self.elements[self.head]
    self.elements[self.head] = nil
    self.head = (self.head % self.capacity) + 1
    self.size = self.size - 1
    return val
end

function CircularQueue:IsEmpty()
    return self.size == 0
end

function CircularQueue:IsFull()
    return self.size == self.capacity
end

function CircularQueue:Peek()
    if self:IsEmpty() then return nil, "queue is empty" end
    return self.elements[self.head]
end

return queue

File: packages/cpp/Queue.cpp
#include "Queue.hpp"

// Queue
void Queue::push(int val) {
    data.push_back(val);
}

void Queue::pop() {
    if (!data.empty()) {
        data.pop_front();
    }
}

int Queue::front() const {
    return data.empty() ? -1 : data.front();
}

bool Queue::isEmpty() const {
    return data.empty();
}

// Deque
void Deque::push_front(int val) {
    data.push_front(val);
}

void Deque::push_back(int val) {
    data.push_back(val);
}

void Deque::pop_front() {
    if (!data.empty()) {
        data.pop_front();
    }
}

void Deque::pop_back() {
    if (!data.empty()) {
        data.pop_back();
    }
}

int Deque::front() const {
    return data.empty() ? -1 : data.front();
}

int Deque::back() const {
    return data.empty() ? -1 : data.back();
}

bool Deque::isEmpty() const {
    return data.empty();
}

CircularQueue::CircularQueue(int k) : head(-1), tail(-1), size(0), capacity(k) {
    data.resize(k);
}

bool CircularQueue::enQueue(int value) {
    if (isFull()) return false;
    if (isEmpty()) head = 0;
    tail = (tail + 1) % capacity;
    data[tail] = value;
    size++;
    return true;
}

bool CircularQueue::deQueue() {
    if (isEmpty()) return false;
    if (head == tail) {
        head = -1;
        tail = -1;
    } else {
        head = (head + 1) % capacity;
    }
    size--;
    return true;
}

int CircularQueue::Front() const {
    return isEmpty() ? -1 : data[head];
}

int CircularQueue::Rear() const {
    return isEmpty() ? -1 : data[tail];
}

bool CircularQueue::isEmpty() const {
    return size == 0;
}

bool CircularQueue::isFull() const {
    return size == capacity;
}
